lib/multi-level.js

```diff
@@ -120,13 +119,16 @@ class MultiLevelCache {
           expiry: Date.now() + this._maxAge
         };
 
-        // Out-of-band cache update
-        // NB: an in-band update would `await` this Promise.all block
-        Promise.all(this._caches.map(cache => cache.set(key, cacheItem))).catch(err => {
-          throw new Error(`Error caching ${key}`, err);
-        }).then(() => {
-          delete this._pendingRefreshes[key];
-        });
+        // Given that we are not ignoring this value, perform an out-of-band cache update
+        const willCache = shouldCache || this.shouldCache;
+        if (typeof willCache === 'function' && willCache(cacheItem.value)) {
+          // NB: an in-band update would `await` this Promise.all block
+          Promise.all(this._caches.map(cache => cache.set(key, cacheItem))).catch(err => {
+            throw new Error(`Error caching ${key}`, err);
+          }).then(() => {
+            delete this._pendingRefreshes[key];
```

Comment:
if you're not caching you need to delete this still, otherwise you're indefinitely caching the promise in here because you got one bad response.

Reply:
Maybe add a test also that doesn't cache the first call, but does cache the second.